### PR TITLE
Add `null` protection

### LINKED ISSUE #2605: [Bug] GitVersion is crashing when trying to update AssemblyInfo.cs files.
**Describe the bug**
Command used in Windows 10 command prompt or PowerShell as well as Azure DevOps build pipeline:
`gitversion /updateassemblyinfo`
GitVersion is crashing when trying to update AssemblyInfo.cs files.

## Expected Behavior
I expected the version in the GitVersion.yml file to be applied to the AssemblyInfo.cs files since it is higher than the latest tag.

## Actual Behavior
```
INFO [02/24/21 12:02:31:25] Working directory: C:\git\Upmc.ServiceBus.Common.V3
INFO [02/24/21 12:02:31:29] Project root is: C:\git\Upmc.ServiceBus.Common.V3\
INFO [02/24/21 12:02:31:29] DotGit directory is: C:\git\Upmc.ServiceBus.Common.V3\.git
INFO [02/24/21 12:02:31:34] Begin: Loading version variables from disk cache
  INFO [02/24/21 12:02:31:34] Begin: Deserializing version variables from cache file C:\git\Upmc.ServiceBus.Common.V3\.git\gitversion_cache\C39A271AF962B211C0A0EB6135B5B939FBB199F0.yml
    WARN [02/24/21 12:02:31:40] Unable to read cache file C:\git\Upmc.ServiceBus.Common.V3\.git\gitversion_cache\C39A271AF962B211C0A0EB6135B5B939FBB199F0.yml, deleting it.
    INFO [02/24/21 12:02:31:41] System.InvalidOperationException: Sequence contains no matching element
   at GitVersion.OutputVariables.VersionVariables.FromFile(String filePath, IFileSystem fileSystem, ILog log)
   at GitVersion.VersionCalculation.Cache.GitVersionCache.LoadVersionVariablesFromDiskCache(GitVersionCacheKey key)
  INFO [02/24/21 12:02:31:41] End: Deserializing version variables from cache file C:\git\Upmc.ServiceBus.Common.V3\.git\gitversion_cache\C39A271AF962B211C0A0EB6135B5B939FBB199F0.yml (Took: 72.49ms)
  INFO [02/24/21 12:02:31:41] End: Loading version variables from disk cache (Took: 73.39ms)
  INFO [02/24/21 12:02:31:45] Using latest commit on specified branch
  INFO [02/24/21 12:02:31:46] Begin: Attempting to inherit branch configuration from parent branch
    INFO [02/24/21 12:02:31:47] Begin: Finding branch source of 'feature/UpdateEventConvention'
      INFO [02/24/21 12:02:31:48] Begin: Finding merge base between 'feature/UpdateEventConvention' and 'develop'.
        INFO [02/24/21 12:02:31:49] Found merge base of e50099d Merged PR 40222: refactored multi-client code
        INFO [02/24/21 12:02:31:49] Merge base of feature/UpdateEventConvention' and 'develop is e50099d Merged PR 40222: refactored multi-client code
      INFO [02/24/21 12:02:31:49] End: Finding merge base between 'feature/UpdateEventConvention' and 'develop'. (Took: 12.31ms)
      INFO [02/24/21 12:02:31:49] Begin: Finding merge base between 'feature/UpdateEventConvention' and 'master'.
      INFO [02/24/21 12:02:31:50] End: Finding merge base between 'feature/UpdateEventConvention' and 'master'. (Took: 4.28ms)
    INFO [02/24/21 12:02:31:50] End: Finding branch source of 'feature/UpdateEventConvention' (Took: 27.41ms)
  INFO [02/24/21 12:02:31:50] End: Attempting to inherit branch configuration from parent branch (Took: 41.27ms)
  ERROR [02/24/21 12:02:31:50] An unexpected error occurred:
System.AggregateException: One or more errors occurred. (Object reference not set to an instance of an object.)
 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at GitVersion.GitObject..ctor(GitObject innerGitObject)
   at GitVersion.Commit..ctor(Commit innerCommit)
   at GitVersion.GitRepository.<>c__DisplayClass27_0.<FindMergeBase>b__0()
   at GitVersion.Helpers.OperationWithExponentialBackoff`2.ExecuteAsync()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at System.Threading.Tasks.Task`1.get_Result()
   at GitVersion.GitRepository.FindMergeBase(ICommit commit, ICommit otherCommit)
   at GitVersion.RepositoryStore.FindMergeBase(IBranch branch, IBranch otherBranch)
   at GitVersion.RepositoryStore.<>c__DisplayClass29_0.<GetMergeCommitsForBranch>b__2(IBranch otherBranch)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.ToArray()
   at System.Linq.OrderedEnumerable`1.ToList()
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at GitVersion.RepositoryStore.GetMergeCommitsForBranch(IBranch branch, Config configuration, IEnumerable`1 excludedBranches)
   at GitVersion.RepositoryStore.FindCommitBranchWasBranchedFrom(IBranch branch, Config configuration, IBranch[] excludedBranches)
   at GitVersion.Configuration.BranchConfigurationCalculator.InheritBranchConfiguration(IBranch targetBranch, BranchConfig branchConfiguration, ICommit currentCommit, Config configuration, IList`1 excludedInheritBranches)
   at GitVersion.Configuration.BranchConfigurationCalculator.GetBranchConfiguration(IBranch targetBranch, ICommit currentCommit, Config configuration, IList`1 excludedInheritBranches)
   at GitVersion.GitVersionContextFactory.Create(GitVersionOptions gitVersionOptions)
   at GitVersion.GitVersionCoreModule.<>c__DisplayClass0_0.<RegisterTypes>b__3()
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.get_Value()
   at GitVersion.VersionCalculation.NextVersionCalculator.get_context()
   at GitVersion.VersionCalculation.NextVersionCalculator.FindVersion()
   at GitVersion.GitVersionCalculateTool.CalculateVersionVariables()
   at GitVersion.GitVersionExecutor.RunGitVersionTool(GitVersionOptions gitVersionOptions)
  INFO [02/24/21 12:02:31:50] Attempting to show the current git graph (please include in issue):
  INFO [02/24/21 12:02:31:50] Showing max of 100 commits
  INFO [02/24/21 12:02:31:64] * d6747b0 4 hours ago  (HEAD -> feature/UpdateEventConvention, origin/feature/UpdateEventConvention)
*   e50099d 3 months ago  (origin/develop, origin/HEAD, develop)
|\
| * 6f79c31 3 months ago
| * d7d38cc 3 months ago
| * 8afc799 3 months ago
| * 4694b49 3 months ago
| * 456f16b 3 months ago
| * aaf1cda 3 months ago
|/
*   57a1600 3 months ago
|\
| * 46fe446 3 months ago
| * 43eccb0 3 months ago
|/
*   dceedd3 3 months ago
|\
| * 68e98d5 3 months ago
| * a498213 3 months ago
| * 6ea2b1c 3 months ago
|/
*   82b1a08 3 months ago
|\
| * b73ff5b 3 months ago
| * a95447f 3 months ago
| * 9354355 3 months ago
| * bd08f2a 3 months ago
| * 3c14cf6 3 months ago
|/
*   7283a86 3 months ago
|\
| * 69a5031 3 months ago
| * d50b889 3 months ago
| * 977f94c 3 months ago
|/
* a2641e5 4 months ago
* 1855878 4 months ago
*   402dc3a 4 months ago
|\
| * 577636d 4 months ago
|/
*   0b317e6 5 months ago
|\
| * 5ac0df6 5 months ago
| *   b67e852 5 months ago
| |\
| |/
|/|
* |   efb2e3d 5 months ago
|\ \
| | * bc0ada2 5 months ago
| |/
| * 8eeb6ea 5 months ago
|/
*   702e36b 7 months ago
|\
| * fc1ef01 7 months ago
|/
*   89522b9 11 months ago
|\
| * 8855b5e 11 months ago
| * ed1858e 11 months ago
|/
*   df60e40 12 months ago
|\
| * 25be4e5 12 months ago
* | 0e243cc 12 months ago
|\|
| * 48a16b8 12 months ago
|/
*   71ce83b 12 months ago
|\
| * c66cc94 12 months ago
* | feeaf17 12 months ago
|\|
| * 40b0f17 12 months ago
| * 3b7d149 12 months ago
| * 55b655b 12 months ago
| * e849bf7 12 months ago
|/
*   b70bc0a 12 months ago
|\
| * 4deb182 12 months ago
| * 23d1ea7 12 months ago
| * 1625e5b 12 months ago
| * 36aeb2c 12 months ago
| * b6c1050 12 months ago
| * 0dbaaba 12 months ago
| * 9372f85 12 months ago
| * f3bdb09 12 months ago
| * 72b5e78 12 months ago
| * cad4f4d 12 months ago
| * f99323a 12 months ago
| * 45f3d8d 12 months ago
| * dd1c115 12 months ago
|/
*   5409029 1 year ago
|\
| * a4a01a9 1 year ago
|/
* 63c548d 1 year, 1 month ago
| * bb73ddf 12 months ago  (origin/master, master)
| * 132ae1b 1 year, 1 month ago  (tag: 2.3.1)
| *   f41353f 1 year, 2 months ago  (tag: 2.3.0)
| |\
| | * f268a2e 1 year, 2 months ago
| |/
|/|
* |   f4a6310 1 year, 2 months ago
|\ \
| * | 8d7e1a4 1 year, 2 months ago
| * | e137d84 1 year, 2 months ago
|/ /
* |   f04e6fe 1 year, 3 months ago
|\ \
| * | d377257 1 year, 3 months ago
|/| |
| * | 6cd047f 1 year, 3 months ago
| * | a93c885 1 year, 3 months ago
|/ /
* |   4a75a59 1 year, 4 months ago
|\ \
| * | 1b8d455 1 year, 4 months ago
| * | cb174f9 1 year, 4 months ago
|/ /
* |   f033008 1 year, 4 months ago
|\ \
| * | 60ea065 1 year, 4 months ago
| * |   7872bdc 1 year, 4 months ago
| |\ \
| |/ /
|/| |
* | |   f515d22 1 year, 5 months ago
|\ \ \
| * \ \   3caad66 1 year, 5 months ago
| |\ \ \
| |/ / /
|/| | |
| * | | 5a2365c 1 year, 5 months ago
| | * | 29f2400 1 year, 4 months ago
| |/ /
|/| |
* | |   a8f1ede 1 year, 5 months ago
|\ \ \
| |/ /
|/| |
| * | 74f7c7b 1 year, 5 months ago
| * | b1616ad 1 year, 5 months ago
|/ /
| *   26b501d 1 year, 6 months ago  (tag: 2.1.1)
| |\
| |/
|/|
* | a929ff5 1 year, 6 months ago
* | cc79df1 1 year, 6 months ago
* | 9c3a568 1 year, 6 months ago
* |   362e029 1 year, 7 months ago
|\ \
| * | 12fd747 1 year, 7 months ago
|/ /
| *   87c7bfd 1 year, 7 months ago  (tag: 2.0.5)
| |\
| | * 3c64ed5 1 year, 7 months ago
| |/
|/|
[GitVersion.yml.txt](https://github.com/GitTools/GitVersion/files/6038051/GitVersion.yml.txt)
```

## Steps to Reproduce
I am not able to share the source code from our internal Git source repository because of intellectual property.

## Context
We use GitVersion as part of our Azure DevOps build pipelines to assign the assembly version information at build time.

## Your Environment
Azure DevOps (hosted)
Classic Build Pipeline

- Version Used: 5.6.6+Branch.main.Sha.80adc3f3b8919b655adca7b796b273fb10478c3f
- Operating System and version (Windows 10 and Azure DevOps):
- Link to your project: (Private)
- Link to your CI build (if appropriate): (Private)

### LINKED ISSUE #2631: [Bug] GitVersion.MsBuild 5.6.7 The command ""C:\Users\UserName\.nuget\packages\gitversion.msbuild\5.6.7\tools\net48/gitversion.exe" "C:\Users\UserName\Documents\source\repos\MyRepo.Test" -output file -outputfile obj\gitversion.json" exited with code 1.
**Describe the bug**
Building a project with **GitVersion.MsBuild 5.6.7** produces:
`The command ""C:\Users\UserName\.nuget\packages\gitversion.msbuild\5.6.7\tools\net48/gitversion.exe" "C:\Users\UserName\Documents\source\repos\MyRepo.Test" -output file -outputfile obj\gitversion.json" exited with code 1.	`

- Version Used:  GitVersion.MsBuild 5.6.7
- Operating System and version (Windows 10, Ubuntu 18.04): Windows 10

Possibly related to: https://github.com/GitTools/GitVersion/pull/2566

### LINKED ISSUE #2695: [Bug] failure when a branch with `--allow-unrelated-histories` is merged in the repository
**Describe the bug**

The following happens when processing a branch in the repository that is merged with a common ancestor for all other branches and the `--allow-unrelated-histories` flag is used. This merged branch does not need to be the one being versioned, only  to be present in the repository.

## Expected Behavior

Gitversion finishes properly.

## Actual Behavior

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at GitVersion.GitObject..ctor(GitObject innerGitObject) in D:\a\GitVersion\GitVersion\src\GitVersion.LibGit2Sharp\Git\GitObject.cs:line 14
   at GitVersion.Commit..ctor(Commit innerCommit) in D:\a\GitVersion\GitVersion\src\GitVersion.LibGit2Sharp\Git\Commit.cs:line 15
   at GitVersion.GitRepository.<>c__DisplayClass27_0.<FindMergeBase>b__0() in D:\a\GitVersion\GitVersion\src\GitVersion.LibGit2Sharp\Git\GitRepository.cs:line 61
   at Polly.Policy`1.<>c__DisplayClass11_0.<Execute>b__0(Context ctx, CancellationToken ct)
   at Polly.Retry.RetryEngine.Implementation[TResult](Func`3 action, Context context, CancellationToken cancellationToken, ExceptionPredicates shouldRetryExceptionPredicates, ResultPredicates`1 shouldRetryResultPredicates, Action`4 onRetry, Int32 permittedRetryCount, IEnumerable`1 sleepDurationsEnumerable, Func`4 sleepDurationProvider)
   at Polly.Retry.RetryPolicy`1.Implementation(Func`3 action, Context context, CancellationToken cancellationToken)
   at Polly.Policy`1.Execute(Func`3 action, Context context, CancellationToken cancellationToken)
   at GitVersion.RepositoryStore.FindMergeBase(IBranch branch, IBranch otherBranch) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Core\RepositoryStore.cs:line 34
   at GitVersion.RepositoryStore.<>c__DisplayClass29_0.<GetMergeCommitsForBranch>b__2(IBranch otherBranch) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Core\RepositoryStore.cs:line 460
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.ToArray()
   at System.Linq.OrderedEnumerable`1.ToList()
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at GitVersion.RepositoryStore.GetMergeCommitsForBranch(IBranch branch, Config configuration, IEnumerable`1 excludedBranches) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Core\RepositoryStore.cs:line 450
   at GitVersion.RepositoryStore.FindCommitBranchWasBranchedFrom(IBranch branch, Config configuration, IBranch[] excludedBranches) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Core\RepositoryStore.cs:line 328
   at GitVersion.Configuration.BranchConfigurationCalculator.InheritBranchConfiguration(IBranch targetBranch, BranchConfig branchConfiguration, ICommit currentCommit, Config configuration, IList`1 excludedInheritBranches) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Configuration\BranchConfigurationCalculator.cs:line 81
   at GitVersion.Configuration.BranchConfigurationCalculator.GetBranchConfiguration(IBranch targetBranch, ICommit currentCommit, Config configuration, IList`1 excludedInheritBranches) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Configuration\BranchConfigurationCalculator.cs:line 47
   at GitVersion.GitVersionContextFactory.Create(GitVersionOptions gitVersionOptions) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Core\GitVersionContextFactory.cs:line 40
   at GitVersion.GitVersionCoreModule.<>c__DisplayClass0_0.<RegisterTypes>b__1() in D:\a\GitVersion\GitVersion\src\GitVersion.Core\GitVersionCoreModule.cs:line 38
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at GitVersion.VersionCalculation.NextVersionCalculator.get_context() in D:\a\GitVersion\GitVersion\src\GitVersion.Core\VersionCalculation\NextVersionCalculator.cs:line 17
   at GitVersion.VersionCalculation.NextVersionCalculator.FindVersion() in D:\a\GitVersion\GitVersion\src\GitVersion.Core\VersionCalculation\NextVersionCalculator.cs:line 32
   at GitVersion.GitVersionCalculateTool.CalculateVersionVariables() in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Core\GitVersionCalculateTool.cs:line 52
   at GitVersion.GitVersionExecutor.RunGitVersionTool(GitVersionOptions gitVersionOptions) in D:\a\GitVersion\GitVersion\src\GitVersion.App\GitVersionExecutor.cs:line 61
```

## Steps to Reproduce
I'm not sure exacty sure how to reproduce this, but this seems to happen when trying to build the version for a branch that dependabot is building.

Here's the [config file used](https://github.com/unoplatform/uno/blob/101f35a2dfe7e5b0242dd76755bd55184133780b/gitversion.yml), if that helps.

## Your Environment
Azure devops windows, mac and linux hosted agents.

### LINKED ISSUE #2734: [Bug] NullReferenceException during dotnet-gitversion execution
This one is a show stopper for the library.  During the build we are getting [NullReferenceException](https://github.com/unitycontainer/container/runs/2852179901?check_suite_focus=true#step:5:124)

```bash
System.NullReferenceException: Object reference not set to an instance of an object.
   at GitVersion.GitObject..ctor(GitObject innerGitObject) in D:\a\GitVersion\GitVersion\src\GitVersion.LibGit2Sharp\Git\GitObject.cs:line 12
   at GitVersion.Commit..ctor(Commit innerCommit) in D:\a\GitVersion\GitVersion\src\GitVersion.LibGit2Sharp\Git\Commit.cs:line 15
   at GitVersion.GitRepository.<>c__DisplayClass27_0.<FindMergeBase>b__0() in D:\a\GitVersion\GitVersion\src\GitVersion.LibGit2Sharp\Git\GitRepository.cs:line 60
   at Polly.Policy`1.<>c__DisplayClass11_0.<Execute>b__0(Context ctx, CancellationToken ct)
   at Polly.Retry.RetryEngine.Implementation[TResult](Func`3 action, Context context, CancellationToken cancellationToken, ExceptionPredicates shouldRetryExceptionPredicates, ResultPredicates`1 shouldRetryResultPredicates, Action`4 onRetry, Int32 permittedRetryCount, IEnumerable`1 sleepDurationsEnumerable, Func`4 sleepDurationProvider)
   at Polly.Retry.RetryPolicy`1.Implementation(Func`3 action, Context context, CancellationToken cancellationToken)
   at Polly.Policy`1.Execute(Func`3 action, Context context, CancellationToken cancellationToken)
   at Polly.Policy`1.Execute(Func`1 action)
   at GitVersion.Helpers.RetryAction`2.Execute(Func`1 operation) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Helpers\RetryAction.cs:line 42
   at GitVersion.GitRepository.FindMergeBase(ICommit commit, ICommit otherCommit) in D:\a\GitVersion\GitVersion\src\GitVersion.LibGit2Sharp\Git\GitRepository.cs:line 58
   at GitVersion.RepositoryStore.FindMergeBase(IBranch branch, IBranch otherBranch) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Core\RepositoryStore.cs:line 46
   at GitVersion.RepositoryStore.<>c__DisplayClass29_0.<GetMergeCommitsForBranch>b__2(IBranch otherBranch) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Core\RepositoryStore.cs:line 466
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.ToArray()
   at System.Linq.OrderedEnumerable`1.ToList()
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at GitVersion.RepositoryStore.GetMergeCommitsForBranch(IBranch branch, Config configuration, IEnumerable`1 excludedBranches) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Core\RepositoryStore.cs:line 450
   at GitVersion.RepositoryStore.FindCommitBranchWasBranchedFrom(IBranch branch, Config configuration, IBranch[] excludedBranches) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Core\RepositoryStore.cs:line 328
   at GitVersion.VersionCalculation.VersionInBranchNameVersionStrategy.GetVersions(String tagPrefixRegex, IBranch currentBranch)+MoveNext() in D:\a\GitVersion\GitVersion\src\GitVersion.Core\VersionCalculation\BaseVersionCalculators\VersionInBranchNameVersionStrategy.cs:line 41
   at GitVersion.VersionCalculation.BaseVersionCalculator.GetBaseVersions(IVersionStrategy strategy)+MoveNext() in D:\a\GitVersion\GitVersion\src\GitVersion.Core\VersionCalculation\BaseVersionCalculator.cs:line 90
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at GitVersion.VersionCalculation.BaseVersionCalculator.GetBaseVersion() in D:\a\GitVersion\GitVersion\src\GitVersion.Core\VersionCalculation\BaseVersionCalculator.cs:line 31
   at GitVersion.VersionCalculation.NextVersionCalculator.FindVersion() in D:\a\GitVersion\GitVersion\src\GitVersion.Core\VersionCalculation\NextVersionCalculator.cs:line 58
   at GitVersion.GitVersionCalculateTool.CalculateVersionVariables() in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Core\GitVersionCalculateTool.cs:line 52
   at GitVersion.GitVersionExecutor.RunGitVersionTool(GitVersionOptions gitVersionOptions) in D:\a\GitVersion\GitVersion\src\GitVersion.App\GitVersionExecutor.cs:line 70
  INFO [06/17/21 19:06:23:48] Attempting to show the current git graph (please include in issue): 
  INFO [06/17/21 19:06:23:48] Showing max of 100 commits
  INFO [06/17/21 19:06:23:58] *   4cbc0d65 in the future  (HEAD -> release/5.12.0, origin/release/5.12.0)
```

Used GitVersion.TooL task when the problem first started, upgraded to GitVersion.MSBuild latest with no success. 
Switched to Actions in hope that will fix the issue, but it is all the same.

This is [the link to failing build](https://github.com/unitycontainer/container/runs/2852179901?check_suite_focus=true#step:5:124). To reproduce, just clone branch locally.

### LINKED ISSUE #2775: [Improvement] More null protection
With #2765 merged we have a good basis to tighten the codebase by adding more `null` guards and allowing for less nullability throughout.

## Detailed Description

One of the worst exceptions to debug is `NullReferenceException`. With `ArgumentNullException` you at least know which argument was `null`, which is a huge improvement. Better yet is to alleviate `null` altogether, or at least move all code dealing with `null` to the perimeter of the codebase. `GitVersion.Core` should preferably be completely free of nullable types.

## Context

We are plagued with `NullReferenceException` surfacing its ugly head every so often. See #2763, #2605, #2695, #2734, #2825 and https://github.com/GitTools/GitVersion/issues/2631#issuecomment-826635079 for examples. We allow `null` by having default, argument-free constructors in classes with fields that then become uninitialized, for instance. We also are missing `null` guards in a lot of constructors that take arguments that should not be allowed to be `null`.

## Possible Implementation

Remove default constructors in classes with fields that need initialization. Add `null` guards throughout the codebase.

### LINKED ISSUE #2825: [Bug] Null Exception
**Describe the bug**
Error: The process '/opt/hostedtoolcache/GitVersion.Tool/5.7.0/x64/dotnet-gitversion' failed with exit code 1

## Expected Behavior
output JSON

## Actual Behavior
Error: The process '/opt/hostedtoolcache/GitVersion.Tool/5.7.0/x64/dotnet-gitversion' failed with exit code 1

## Context
I can't use gitversion in my CI/CD

## Your Environment

- Version Used: 5.x - latest
- Operating System and version (Windows 10, Ubuntu 18.04): Ubuntu 20.04
- Link to your project:
- Link to your CI build (if appropriate):

```
  ERROR [08/27/21 3:34:09:55] An unexpected error occurred:
System.NullReferenceException: Object reference not set to an instance of an object.
   at GitVersion.GitObject..ctor(GitObject innerGitObject) in D:\a\GitVersion\GitVersion\src\GitVersion.LibGit2Sharp\Git\GitObject.cs:line 14
   at GitVersion.Commit..ctor(Commit innerCommit) in D:\a\GitVersion\GitVersion\src\GitVersion.LibGit2Sharp\Git\Commit.cs:line 15
   at GitVersion.GitRepository.<>c__DisplayClass27_0.<FindMergeBase>b__0() in D:\a\GitVersion\GitVersion\src\GitVersion.LibGit2Sharp\Git\GitRepository.cs:line 64
   at Polly.Policy`1.<>c__DisplayClass11_0.<Execute>b__0(Context ctx, CancellationToken ct)
   at Polly.Retry.RetryEngine.Implementation[TResult](Func`3 action, Context context, CancellationToken cancellationToken, ExceptionPredicates shouldRetryExceptionPredicates, ResultPredicates`1 shouldRetryResultPredicates, Action`4 onRetry, Int32 permittedRetryCount, IEnumerable`1 sleepDurationsEnumerable, Func`4 sleepDurationProvider)
   at Polly.Retry.RetryPolicy`1.Implementation(Func`3 action, Context context, CancellationToken cancellationToken)
   at Polly.Policy`1.Execute(Func`3 action, Context context, CancellationToken cancellationToken)
   at Polly.Policy`1.Execute(Func`1 action)
   at GitVersion.Helpers.RetryAction`2.Execute(Func`1 operation) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Helpers\RetryAction.cs:line 37
   at GitVersion.GitRepository.FindMergeBase(ICommit commit, ICommit otherCommit) in D:\a\GitVersion\GitVersion\src\GitVersion.LibGit2Sharp\Git\GitRepository.cs:line 62
   at GitVersion.RepositoryStore.FindMergeBase(IBranch branch, IBranch otherBranch) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Core\RepositoryStore.cs:line 48
   at GitVersion.RepositoryStore.<>c__DisplayClass29_0.<GetMergeCommitsForBranch>b__2(IBranch otherBranch) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Core\RepositoryStore.cs:line 454
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.ToArray()
   at System.Linq.OrderedEnumerable`1.ToList()
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at GitVersion.RepositoryStore.GetMergeCommitsForBranch(IBranch branch, Config configuration, IEnumerable`1 excludedBranches) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Core\RepositoryStore.cs:line 438
   at GitVersion.RepositoryStore.FindCommitBranchWasBranchedFrom(IBranch branch, Config configuration, IBranch[] excludedBranches) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Core\RepositoryStore.cs:line 321
   at GitVersion.Configuration.BranchConfigurationCalculator.InheritBranchConfiguration(IBranch targetBranch, BranchConfig branchConfiguration, ICommit currentCommit, Config configuration, IList`1 excludedInheritBranches) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Configuration\BranchConfigurationCalculator.cs:line 81
   at GitVersion.Configuration.BranchConfigurationCalculator.GetBranchConfiguration(IBranch targetBranch, ICommit currentCommit, Config configuration, IList`1 excludedInheritBranches) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Configuration\BranchConfigurationCalculator.cs:line 47
   at GitVersion.GitVersionContextFactory.Create(GitVersionOptions gitVersionOptions) in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Core\GitVersionContextFactory.cs:line 40
   at GitVersion.GitVersionCoreModule.<>c__DisplayClass0_0.<RegisterTypes>b__1() in D:\a\GitVersion\GitVersion\src\GitVersion.Core\GitVersionCoreModule.cs:line 38
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at GitVersion.VersionCalculation.NextVersionCalculator.get_context() in D:\a\GitVersion\GitVersion\src\GitVersion.Core\VersionCalculation\NextVersionCalculator.cs:line 17
   at GitVersion.VersionCalculation.NextVersionCalculator.FindVersion() in D:\a\GitVersion\GitVersion\src\GitVersion.Core\VersionCalculation\NextVersionCalculator.cs:line 32
   at GitVersion.GitVersionCalculateTool.CalculateVersionVariables() in D:\a\GitVersion\GitVersion\src\GitVersion.Core\Core\GitVersionCalculateTool.cs:line 52
   at GitVersion.GitVersionExecutor.RunGitVersionTool(GitVersionOptions gitVersionOptions) in D:\a\GitVersion\GitVersion\src\GitVersion.App\GitVersionExecutor.cs:line 70
  INFO [08/27/21 3:34:09:55] Attempting to show the current git graph (please include in issue): 
  INFO [08/27/21 3:34:09:55] Showing max of 100 commits
  INFO [08/27/21 3:34:09:61] * e94028f 26 minutes ago  (HEAD -> build, origin/master, master)
```

### PR DESCRIPTION
**Pull request by asbjornu**
Add `null` guards to `GitObject` and `ObjectId` to avoid `NullReferenceException`.

## Description

This PR is an exploration of which `null` guards we can add to avoid `NullReferenceException` and possibly also avoiding `null` before they reach `GitVersion.Core`.

## Related Issue
Resolves #2775.

The following issues aren't fixed as this PR doesn't address the underlying issues, but this PR will provide a much more informative stack trace so if anyone encounters the underlying issues again, we'll have a better chance at figuring out what the issue is and actually fix it.

Closes #2605.
Closes #2631.
Closes #2695.
Closes #2734.
Closes #2825.

## Motivation and Context
See #2775 and all linked issues.

## How Has This Been Tested?
I'm counting on our existing test coverage to sufficiently exercise the code involved.

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/GitVersion.LibGit2Sharp/Git/GitObject.cs [hunk 1]
```diff
@@ -9,6 +9,9 @@ internal class GitObject : IGitObject
 
     internal GitObject(LibGit2Sharp.GitObject innerGitObject)
     {
+        if (innerGitObject == null)
+            throw new ArgumentNullException(nameof(innerGitObject));
+
```
Comment (arturcic) [inline src/GitVersion.LibGit2Sharp/Git/GitObject.cs:14]:
```suggestion
        innerGitObject = innerGitObject.NotNull();
```

Comment (asbjornu) [reply src/GitVersion.LibGit2Sharp/Git/GitObject.cs:14]: Ah, but of course.

**Comment by arturcic** [comment]
Can you also add the null check for these classes? `BranchCollection, Reference, TagCollection`

**Review by arturcic — APPROVED** [review]
LGTM

**Comment by arturcic** [comment]
@asbjornu will you also add for the other 2?

**Comment by asbjornu** [comment]
@arturcic, yes, that's the plan. I just want every commit to be green so I'm confident that none of the added null checks break anything – and if they do, I want to know which one it is. But I'm unable to get a successful build out of 2790f277c1c516f627764a1894b5b17c27670f18 due to some docker login failing. Ideas?

**Comment by asbjornu** [comment]
Yea, I just want this ❌ green. 😅 

image

I'll amend and force push, hopefully it turns green then.

**Comment by arturcic** [comment]
I think it's failing on your branch on your repo, not on GtiTools, because it's trying to use some envvars that you do not have setup

**Comment by asbjornu** [comment]
Yes, that was my thought as well, @arturcic. But that doesn't explain why previous commits as well as the amended commit are all green.

successful commit

Anyway, as the commit is now successful, I'll continue pushing the other files as well.

**Pull request merged**